src/meta/src/stream/scale.rs

```diff
@@ -0,0 +1,827 @@
+            for actor in &fragment.actors {
+                if actors_to_remove.contains_key(&actor.actor_id) {
+                    continue;
+                }
+
+                let parallel_unit_id = actor_status
+                    .get(&actor.actor_id)
+                    .unwrap()
+                    .parallel_unit
+                    .as_ref()
+                    .unwrap()
+                    .id;
+                new_actor_ids.insert(
+                    actor.actor_id as ActorId,
+                    parallel_unit_id as ParallelUnitId,
+                );
+            }
+
+            for (actor_id, parallel_unit_id) in actors_to_create {
+                new_actor_ids.insert(*actor_id, *parallel_unit_id as ParallelUnitId);
+            }
+
+            fragment_actors.insert(*fragment_id, new_actor_ids);
+        }
+
+        // Generate fragment reschedule plan
+        let mut reschedule_fragment: HashMap<FragmentId, Reschedule> =
+            HashMap::with_capacity(reschedule.len());
+        for (
+            fragment_id,
+            ParallelUnitReschedule {
+                added_parallel_units,
+                removed_parallel_units,
+            },
+        ) in reschedule
+        {
+            assert_eq!(added_parallel_units.len(), removed_parallel_units.len());
+
+            let replace_parallel_unit_map: BTreeMap<_, _> = removed_parallel_units
+                .into_iter()
+                .zip_eq(added_parallel_units.into_iter())
+                .collect();
+
+            let actors_to_create = fragment_actors_to_create
+                .get(&fragment_id)
+                .cloned()
+                .unwrap_or_default()
+                .into_keys()
+                .collect();
+            let actors_to_remove = fragment_actors_to_remove
+                .get(&fragment_id)
+                .cloned()
+                .unwrap_or_default()
+                .into_keys()
+                .collect();
+
+            let fragment_parallel_unit_to_actor: BTreeMap<_, _> = fragment_actors
+                .get(&fragment_id)
+                .unwrap()
+                .iter()
+                .map(|(actor_id, parallel_unit_id)| {
+                    (*parallel_unit_id as ParallelUnitId, *actor_id as ActorId)
+                })
+                .collect();
+
+            let fragment = fragment_map.get(&fragment_id).unwrap();
+
+            assert_eq!(fragment.distribution_type(), FragmentDistributionType::Hash);
+
+            assert!(
+                !fragment_parallel_unit_to_actor.is_empty(),
+                "hash dispatcher should have at least one downstream actor"
+            );
+
+            let upstream_dispatcher_mapping = if fragment_parallel_unit_to_actor.len() == 1 {
+                ActorMapping {
+                    original_indices: vec![VIRTUAL_NODE_COUNT as u64 - 1],
+                    data: vec![*fragment_parallel_unit_to_actor.first_key_value().unwrap().1],
+                }
```

Comment:
For simple dispatcher, the `mapping` is ignored. As we won't change the parallism of a fragment when migrating, we can assume this will only occur for simple dispatcher.